doc/developer/design/20241008_expression_cache.md

```diff
@@ -0,0 +1,183 @@
+object is currently deployed in a cluster. This cache contains the expressions that are deployed in
+a cluster, but not necessarily the expressions that would result from optimization from the current
+catalog contents.
+
+### Cache API
+
+Below is the API that the cache will present. It may be further wrapped with typed methods that
+take care of serializing and deserializing bytes. Additionally, we probably don't need a trait when
+implementing.
+
+```Rust
+
+
+/// All the cached expressions for a single `GlobalId`.
+///
+/// Note: This is just a placeholder for now, don't index too hard on the exact fields. I haven't
+/// done the necessary research to figure out what they are.
+struct Expressions {
+    local_mir: OptimizedMirRelationExpr,
+    global_mir: DataflowDescription<OptimizedMirRelationExpr>,
+    physical_plan: DataflowDescription<mz_compute_types::plan::Plan>,
+    dataflow_metainfos: DataflowMetainfo<Arc<OptimizerNotice>>,
+    notices: SmallVec<[Arc<OptimizerNotice>; 4]>,
+}
+
+struct ExpressionCache {
+    deploy_generation: u64,
+    information_needed_to_connect_to_durable_store: _,
+}
+
+impl ExpressionCache {
+    /// Creates a new [`ExpressionCache`] for `deploy_generation`.
+    fn new(&mut self, deploy_generation: u64, information_needed_to_connect_to_durable_store: _) -> Self;
+
+    /// Remove all entries in current deploy generation that depend on a global ID that is not
+    /// present in `current_ids`.
+    ///
+    /// If `remove_prior_gens` is `true`, all previous generations are durably  removed from the
+    /// cache.
+    ///
+    /// Returns all cached expressions.
+    fn open(&mut self, current_ids: &BTreeSet<GlobalId>, remove_prior_gens: bool) -> Vec<(GlobalId, Expressions)>;
+
+    /// Durably inserts `expressions` into current deploy generation.
+    ///
+    /// Returns a [`Future`] that completes once `expressions` have been made durable.
+    ///
+    /// Panics if any `GlobalId` already exists in the cache.
+    fn insert_expressions(&mut self, expressions: Vec<(GlobalId, Expressions)>) -> impl Future<Output = ()>;
+
+    /// Durably remove and return all entries in current deploy generation that depend on an ID in
+    /// `dropped_ids` .
+    ///
+    /// Optional for v1.
+    fn invalidate_entries(&mut self, dropped_ids: BTreeSet<GlobalId>) -> Vec<(GlobalId, Expressions)>;
+}
+```
+
+### Startup
+
+Below is a detailed set of steps that will happen in startup.
+
+1. Call `ExpressionCache::open` to remove any invalid entries and retrieve cached entries. When
+   passing in the arguments, `remove_prior_gens == !read_only_mode`.
+2. While opening the catalog, for each object:
+    a. If the object is present in the cache, use cached optimized expression.
+    b. Else generate the optimized expressions and insert the expressions via
+       `ExpressionCache::insert_expressions`.
+
+### DDL - Create
+
+1. Execute catalog transaction.
+2. Update cache via `ExpressionCache::insert_expressions`.
```

Comment:
Unfortunately, adding new indexes can also invalidate cached plans, when the new index should be picked up by a plan.

Therefore, when a new index is created on an object `o1`, I'd suggest to invalidate all plans of objects `o2` where `o2` depends on `o1`. It's enough to go after only direct compute dependents, i.e., no need to transitively traverse dependencies. (But non-compute objects, i.e. views, should be transitively traversed.)

Reply:
One more invalidation issue: There are feature flags affecting the optimizer, see `optimizer_feature_flags!`. When any of these change, I don't see any other thing to do than just to throw away the entire cache, unfortunately.

Reply:
(The invalidation on drop is critical, because otherwise we'd crash due to wanting to use a plan that refers to a dropped object. However, for the above extra invalidations, it would be also a possibility to just not do them: they are less critical, in that the plan wouldn't cause a crash. But I think we should still do them, otherwise the behavior of the system would become a bit complicated to explain.)

Reply:
> Unfortunately, adding new indexes can also invalidate cached plans, when the new index should be picked up by a plan.

Technically this is only true after a restart, the new index will only be used by the plan after envd has restarted. For a given deploy generation, the expectation is that envd will restart 0 times, so I think it's _probably_ fine to change the behavior s.t. dataflows are not replanned after a restart within a deploy generation. The only real downside here is that we lose the ability to force a plan to use a new index by restarting the env. Though I'm not sure if that's something we've ever done or plan on doing. With the current design, every new deploy generation will replan all objects, so the new index will get picked up then.

> One more invalidation issue: There are feature flags affecting the optimizer, see optimizer_feature_flags!. When any of these change, I don't see any other thing to do than just to throw away the entire cache, unfortunately.

Good call out. I think what we can do here is store a copy of the flags used to optimize the expressions in the cache. On restart during reconciliation we can compare the current flags with the saved flags and delete everything if they don't match.

> The invalidation on drop is critical, because otherwise we'd crash due to wanting to use a plan that refers to a dropped object.

I don't think we need to immediately invalidate expressions on drop. We can wait for a restart and reconcile the cache with the existing objects in the catalog, and delete cached expressions that refer to deleted objects.

Reply:
>> The invalidation on drop is critical, because otherwise we'd crash due to wanting to use a plan that refers to a dropped object.

> I don't think we need to immediately invalidate expressions on drop. We can wait for a restart and reconcile the cache with the existing objects in the catalog, and delete cached expressions that refer to deleted objects.

Sorry, yes, I just meant that it has to be invalidated after a drop before bootstrap would use them. Whether that happens immediately at a drop or only during a bootstrapping reconciliation doesn't matter, both options seem fine to me.

>> Unfortunately, adding new indexes can also invalidate cached plans, when the new index should be picked up by a plan.

> Technically this is only true after a restart, the new index will only be used by the plan after envd has restarted. For a given deploy generation, the expectation is that envd will restart 0 times, so I think it's probably fine to change the behavior s.t. dataflows are not replanned after a restart within a deploy generation. The only real downside here is that we lose the ability to force a plan to use a new index by restarting the env. Though I'm not sure if that's something we've ever done or plan on doing. With the current design, every new deploy generation will replan all objects, so the new index will get picked up then.

Well, yes, ignoring index creations could be kind of ok, but still, a bit unpleasant. For example, we'd have an asymmetric behavior between index drops and creations: drops would take effect at any restart, but creations would take effect only at the next deploy generation. This way, the cache will be one more thing that one has to keep in mind when trying to figure out what indexes will be used when. On the other hand, if we did invalidation on creation, then the caching would be completely transparent.

>> One more invalidation issue: There are feature flags affecting the optimizer, see optimizer_feature_flags!. When any of these change, I don't see any other thing to do than just to throw away the entire cache, unfortunately.

> Good call out. I think what we can do here is store a copy of the flags used to optimize the expressions in the cache. On restart during reconciliation we can compare the current flags with the saved flags and delete everything if they don't match.

Yes, that seems like a good solution for this! Specifically, I'd recommend storing the `OptimizerFeatures` object.

Reply:
>> One more invalidation issue: There are feature flags affecting the optimizer, see optimizer_feature_flags!. When any of these change, I don't see any other thing to do than just to throw away the entire cache, unfortunately.
>
> Good call out. I think what we can do here is store a copy of the flags used to optimize the expressions in the cache. On restart during reconciliation we can compare the current flags with the saved flags and delete everything if they don't match.

I've added some details around this.

Reply:
> Well, yes, ignoring index creations could be kind of ok, but still, a bit unpleasant. For example, we'd have an asymmetric behavior between index drops and creations: drops would take effect at any restart, but creations would take effect only at the next deploy generation. This way, the cache will be one more thing that one has to keep in mind when trying to figure out what indexes will be used when. On the other hand, if we did invalidation on creation, then the caching would be completely transparent.

Hmm, that's a good point. Unfortunately this adds a new complexity. Inserting/removing from the cache is a best effort, so we need to be able to reconcile the cache on startup. If we crash before invalidating the cache as a result of an index creation, then we need to somehow detect this on startup and reconcile the cache state.

My current idea for this is the following. During the reconciliation phase we get the set of all indexes that exist in the cache (`cache_i`) and a set of all indexes that exist in the catalog (`catalog_i`). We then compute the difference, `catalog_i - cache_i`, which is the set of new indexes that didn't make it into the cache. We can then go through the regular invalidation that we would have done when those new indexes were created.

Reply:
Sounds good, thank you!

Reply:
Agreed, that sounds good to me as well.

Reply:
Ok, I've updated the doc to include this logic.